torch/_inductor/utils.py

```diff
@@ -1812,9 +1812,15 @@ def tensor_is_aligned(tensor: torch.Tensor):
     # but symbolic storage_offsets are. For consistency, we suppress guard creation
     # upon performing this check: that ensures that we don't add recompiles when we
     # add this logic.
-    return (
-        tensor.storage_offset() * get_dtype_size(tensor.dtype)
-    ) % GPU_ALIGN_BYTES == 0
+    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+
+    try:
+        return guard_size_oblivious(
+            (tensor.storage_offset() * get_dtype_size(tensor.dtype)) % GPU_ALIGN_BYTES
+            == 0
+        )
+    except torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode:
+        return False
```

Comment:
Don't catch the exception. Instead, more directly do something like statically_known_true

Reply:
Just want to make sure my understanding is correct:  For this case, since we can not evaluate the expr in a static way, so statically_known_true is just to return False, but during runtime, statically_known_true will use the real values to evaluate. 

If so, let's assume that during runtime, the value returned by statically_known_true is True, what will happen? Since 
during compilation,  statically_known_true is False, so the compiled graph is based on "statically_known_true is False", but during runtime, statically_known_true becomes True, then do we need to recompile?

Reply:
@ezyang any idea?

Reply:
I mean, this is a context dependent question. When you are testing this to determine if you can apply an optional optimization, such a difference would be fine.

Reply:
FWIW, your code here is also doing the same thing

Reply:
> I mean, this is a context dependent question. When you are testing this to determine if you can apply an optional optimization, such a difference would be fine.

A bit confused here, do you still need to recompile or not for the case I described?

Reply:
> FWIW, your code here is also doing the same thing

Which code? Could you elaborate a bit more?

Reply:
You don't necessarily have to recompile.

Let me give an example. Suppose you are compiling 

```
def f(x: int, y: int):
    return x + (y % 2)
```

You are considering an optimization to optimize this into:

```
def f(x: int, y: int):
    return x
```

Clearly, this optimization is only valid when y % 2 == 0. So you want to test if this expression is true. If it is statically known to be true, the optimization is safe. But suppose we do not know. It is safe not to optimize the program and keep the original, because the original program is valid for both cases.

Reply:
I see. Just want to make sure I understand correct:
```
def f(x: int, y: int):
    if y % 2 == 0:
       return x + 1
    else:
       return x
```
Assume `y` is an unbacked symint, so dynamo throws an error. Then we put `statically_known_true` around the statement as below
```
def f(x: int, y: int):
    if statically_known_true(y % 2 == 0):
       return x + 1
    else:
       return x
```
if dynamo can figure out the value of the statement by static analysis, let's say y is even, then codes will be simplified as:
```
def f(x: int, y: int):
       return x + 1
```
if it can not figure out the value, then dynamo keep it as it is:

```
def f(x: int, y: int):
    if y % 2 == 0:
       return x + 1
    else:
       return x
```
Is that correct?

Reply:
No, not really. You're not "put[ting] statically_known_true around the statement as below", this is happening entirely in the optimizer.

Reply:
> No, not really. You're not "put[ting] statically_known_true around the statement as below", this is happening entirely in the optimizer.

Hmm, did not get it. In this PR we use `statically_known_true` to resolve the issue, why in the above example, we don't use `statically_known_true`, is it because these are user codes not inside inductor?

Is `statically_known_true(x)` to give a hint to the compiler that `x` is True, or it asks the compiler to do static analysis on `x` to determine whether it is true or not?

Reply:
You're mixing up user code (the code sample we were looking at) and Inductor lowering code (where the statically_known_true shows up)

Reply:
@ezyang I see, but what does `statically_known_true(x)` actually do?  Let's assume that if inductor wants to apply an optimization, it sees `statically_known_true(x)`. Does it tell inductor that "try your best to do static analysis about x, but if you are uncertain about x's value, then don't apply the optimization"?

Reply:
you are saying things that individually are right but they are not being put together correctly. If you have a symbolic expression, you use statically_known_true in Inductor to do this analysis. Then you use the result to make a decision about whether or not to do an optimization or not

Reply:
@ezyang let's assume that during compile time `statically_known_true(x)` returns False, and compiler decides not to do optimization, but during run-time, `statically_known_true(x)` returns True, will compiler apply the optimization, If it will, then it seems like that the compile's decision at compile time does not matter which could be changed later during run-time, is that the case?

Reply:
That's not how it works. The statically_known_true test is a compile time concept. At runtime we never call "statically_known_true".

Reply:
@ezyang I see, if we only do it during compile time, does that mean that if we can not reason the value of `x` during compile time, we will not do optimization. But `x` could be True during runtime, so it misses the optimization opportunity.

Reply:
I mean, it's debatable whether or not there was an "optimization opportunity" at all in the first place. Is changing "if x == 2 then x * 2 else x + 5" to "x * 2" and optimization? Not really, it's just wrong LOL